### PR TITLE
[Enhancement] Add FE memory related metrics

### PR DESCRIPTION
**Pull request by gengjun-git**
1. Add some object count metrics to help debug memory problems.
2. For object whose properties are almost basic type, we use SizeEstimator to calculate its bytes, like Tablet, Replica.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 27 / 120 (22.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|0|1|00.00%|[585]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|0|9|00.00%|[819, 820, 822, 823, 824, 825, 826, 827, 828]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|21|103|20.39%|[522, 530, 531, 533, 546, 558, 559, 563, 564, 565, 566, 568, 569, 573, 574, 575, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 591, 592, 593, 594, 595, 597, 598, 599, 600, 601, 604, 605, 606, 608, 609, 610, 611, 612, 614, 615, 616, 619, 621, 622, 625, 626, 627, 631, 632, 636, 637, 641, 644, 645, 646, 649, 650, 651, 654, 657, 658, 662, 663, 664, 665, 666, 669, 670, 674, 675, 685, 686, 687, 688]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|1|2|50.00%|[359]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|2|2|100.00%|[]|

**Comment by nshangyiming** [comment]
@HangyuanLiu plz also take a look at this pr

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 27 / 120 (22.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|0|1|00.00%|[585]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|0|9|00.00%|[819, 820, 822, 823, 824, 825, 826, 827, 828]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|21|103|20.39%|[522, 530, 531, 533, 546, 558, 559, 563, 564, 565, 566, 568, 569, 573, 574, 575, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 591, 592, 593, 594, 595, 597, 598, 599, 600, 601, 604, 605, 606, 608, 609, 610, 611, 612, 614, 615, 616, 619, 621, 622, 625, 626, 627, 631, 632, 636, 637, 641, 644, 645, 646, 649, 650, 651, 654, 657, 658, 662, 663, 664, 665, 666, 669, 670, 674, 675, 685, 686, 687, 688]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|1|2|50.00%|[359]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 76 / 113 (67.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|0|1|00.00%|[601]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|0|1|00.00%|[54]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|0|1|00.00%|[915]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|0|3|00.00%|[354, 356, 358]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|0|4|00.00%|[819, 823, 825, 827]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|0|3|00.00%|[243, 247, 251]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|0|1|00.00%|[208]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|0|2|00.00%|[817, 821]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|70|90|77.78%|[521, 531, 532, 542, 552, 553, 563, 573, 574, 584, 594, 604, 614, 624, 634, 644, 654, 664, 674, 684]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 76 / 113 (67.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|0|1|00.00%|[601]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|0|1|00.00%|[54]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|0|1|00.00%|[915]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|0|3|00.00%|[354, 356, 358]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|0|4|00.00%|[819, 823, 825, 827]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|0|3|00.00%|[243, 247, 251]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|0|1|00.00%|[208]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|0|2|00.00%|[817, 821]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|70|90|77.78%|[521, 531, 532, 542, 552, 553, 563, 573, 574, 584, 594, 604, 614, 624, 634, 644, 654, 664, 674, 684]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

File: test/sql/test_metrics/R/test_memory_metrics [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+1
+-- !result
+-- name: test_memory_metrics_replica_bytes
+shell: curl '${url}/metrics' | grep starrocks_fe_memory | grep replica_bytes | wc -l
+-- result:
+0
+1
+-- !result
+-- name: test_memory_metrics_tablet_bytes
+shell: curl '${url}/metrics' | grep starrocks_fe_memory | grep tablet_bytes | wc -l
```
Comment (nshangyiming) [inline test/sql/test_metrics/R/test_memory_metrics:62]: should we create some tables to check if these metrics are correct?

File: fe/fe-core/src/main/java/com/starrocks/metric/MetricRepo.java [hunk 2]
```diff
@@ -507,6 +513,181 @@ public Long getValue() {
...
+        };
+        tabletCnt.addLabel(new MetricLabel("type", "tablet_count"));
+        STARROCKS_METRIC_REGISTER.addMetric(tabletCnt);
+
+        GaugeMetric<Long> tabletBytes = new GaugeMetric<Long>("memory", MetricUnit.BYTES,
+                "The bytes of tablets") {
+            @Override
+            public Long getValue() {
+                return GlobalStateMgr.getCurrentInvertedIndex().getTabletCount()
+                        * SizeEstimator.estimate(new LocalTablet());
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/metric/MetricRepo.java:532]: will this count the ref to `private List<Replica> replicas;` member?

Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/metric/MetricRepo.java:532]: Yes, but for the `new LocalTablet()`, the repicas is null, so only the basic types will be included.

File: test/sql/test_metrics/R/test_memory_metrics (hunk 1 above)
Comment (gengjun-git) [reply test/sql/test_metrics/R/test_memory_metrics:62]: The cluster can create many tables, it's not easy to calculate the number.

**Comment by gengjun-git** [comment]
@megifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 113 / 113 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|90|90|100.00%|[]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

File: test/sql/test_metrics/R/test_memory_metrics (hunk 1 above)
Comment (nshangyiming) [reply test/sql/test_metrics/R/test_memory_metrics:62]: we don't need a precise number, just  check `> xxx`

Comment (nshangyiming) [reply test/sql/test_metrics/R/test_memory_metrics:62]: does sql test support > xxx ?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 113 / 113 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|90|90|100.00%|[]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

**Comment by gengjun-git** [comment]
@mergifyio rebase

**Review by nshangyiming — DISMISSED** [review]

**Review by Astralidea — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 109 / 141 (77.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|0|1|00.00%|[214]|
|:large_blue_circle:|com/starrocks/qe/QueryDetailQueue.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/common/util/ProfileManager.java|0|6|00.00%|[269, 271, 273, 278, 280, 282]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|86|110|78.18%|[524, 534, 535, 545, 555, 556, 566, 576, 577, 587, 597, 607, 617, 627, 637, 647, 657, 667, 677, 687, 697, 707, 717, 727]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 141 / 141 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/QeProcessorImpl.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TxnStateCallbackFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/QueryDetailQueue.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|110|110|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/util/ProfileManager.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/lake/compaction/CompactionMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunHistory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobMgr.java|3|3|100.00%|[]|

**Review by nshangyiming — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**